File: CRM-Client/src/api/calendar.ts
import request from '@/utils/request'

export interface TodoCount {
  lead: number
  customer: number
  opportunity: number
  payment: number
  total: number
}

export interface CalendarMonthResponse {
  year: number
  month: number
  todos: Record<string, TodoCount>
}

export interface LeadFollowUpTodo {
  id: number
  lead_id: number
  lead_name: string
  contact_name: string
  contact_phone: string
  next_action: string | null
  next_follow_time: string | null
  is_overdue: boolean
}

export interface CustomerFollowUpTodo {
  id: number
  customer_id: number
  account_name: string
  next_action: string | null
  next_follow_time: string | null
  is_overdue: boolean
}

export interface OpportunityTodo {
  id: number
  opportunity_name: string
  customer_name: string
  total_amount: number
  expected_closing_date: string
  current_stage_name: string | null
  is_overdue: boolean
}

export interface PaymentPlanTodo {
  id: number
  contract_id: number
  contract_name: string
  customer_name: string
  stage_name: string
  planned_amount: number
  due_date: string
  is_overdue: boolean
}

export interface CalendarDateDetailResponse {
  date: string
  lead_follow_ups: LeadFollowUpTodo[]
  customer_follow_ups: CustomerFollowUpTodo[]
  opportunities: OpportunityTodo[]
  payment_plans: PaymentPlanTodo[]
  total_count: number
}

// FollowUp AI 跟进相关类型
export interface FollowUpParseRequest {
  todo_type: string
  todo_id: number
  user_input: string
}

export interface ParsedAction {
  skill: string
  action: string
  params: Record<string, any>
}

export interface FollowUpParseResponse {
  actions: ParsedAction[]
  reply_text: string
  context: TodoContextResponse
}

export interface FollowUpExecuteRequest {
  actions: ParsedAction[]
}

export interface FollowUpExecuteResult {
  success: boolean
  message: string
  data?: Record<string, any>
}

export interface FollowUpExecuteResponse {
  success: boolean
  message: string
  results: FollowUpExecuteResult[]
}

export interface TodoContextResponse {
  todo_type: string
  todo_id: number
  entity_type: string
  entity_id: number
  entity_info: Record<string, any>
  current_next_follow_time: string | null
  current_next_action: string | null
}

export const calendarApi = {
  getMonthTodos: (year: number, month: number) => {
    return request.get<CalendarMonthResponse>('/v1/calendar/todos', {
      params: { year, month }
    })
  },

  getDateTodos: (date: string) => {
    return request.get<CalendarDateDetailResponse>('/v1/calendar/todos/date', {
      params: { date }
    })
  },

  getTodoContext: (todoType: string, todoId: number) => {
    return request.get<TodoContextResponse>(`/v1/calendar/todos/${todoType}/${todoId}/context`)
  },

  parseFollowUp: (data: FollowUpParseRequest) => {
    return request.post<FollowUpParseResponse>('/v1/calendar/follow-up/parse', data)
  },

  executeFollowUp: (data: FollowUpExecuteRequest) => {
    return request.post<FollowUpExecuteResponse>('/v1/calendar/follow-up/execute', data)
  }
}

File: CRM-Server/app/schemas/calendar.py
from pydantic import BaseModel
from typing import List, Optional
from datetime import date, datetime


class TodoCount(BaseModel):
    lead: int = 0
    customer: int = 0
    opportunity: int = 0
    payment: int = 0
    total: int = 0


class CalendarMonthResponse(BaseModel):
    year: int
    month: int
    todos: dict[str, TodoCount]


class LeadFollowUpTodo(BaseModel):
    id: int
    lead_id: int
    lead_name: str
    contact_name: str
    contact_phone: str
    next_action: Optional[str]
    next_follow_time: Optional[datetime]
    is_overdue: bool


class CustomerFollowUpTodo(BaseModel):
    id: int
    customer_id: int
    account_name: str
    next_action: Optional[str]
    next_follow_time: Optional[datetime]
    is_overdue: bool


class OpportunityTodo(BaseModel):
    id: int
    opportunity_name: str
    customer_name: str
    total_amount: float
    expected_closing_date: date
    current_stage_name: Optional[str]
    is_overdue: bool


class PaymentPlanTodo(BaseModel):
    id: int
    contract_id: int
    contract_name: str
    customer_name: str
    stage_name: str
    planned_amount: float
    due_date: date
    is_overdue: bool


class CalendarDateDetailResponse(BaseModel):
    date: str
    lead_follow_ups: List[LeadFollowUpTodo]
    customer_follow_ups: List[CustomerFollowUpTodo]
    opportunities: List[OpportunityTodo]
    payment_plans: List[PaymentPlanTodo]
    total_count: int


# FollowUp AI 跟进相关 Schema
from typing import Dict, Any


class FollowUpParseRequest(BaseModel):
    """跟进解析请求"""
    todo_type: str
    todo_id: int
    user_input: str


class FollowUpParseResponse(BaseModel):
    """跟进解析响应"""
    actions: List[Dict[str, Any]]
    reply_text: str
    context: Dict[str, Any]


class FollowUpExecuteRequest(BaseModel):
    """跟进执行请求"""
    actions: List[Dict[str, Any]]


class FollowUpExecuteResult(BaseModel):
    """单个 action 执行结果"""
    success: bool
    message: str
    data: Optional[Dict[str, Any]] = None


class FollowUpExecuteResponse(BaseModel):
    """跟进执行响应"""
    success: bool
    message: str
    results: List[FollowUpExecuteResult]


class TodoContextResponse(BaseModel):
    """待办上下文响应"""
    todo_type: str
    todo_id: int
    entity_type: str
    entity_id: int
    entity_info: Dict[str, Any]
    current_next_follow_time: Optional[str]
    current_next_action: Optional[str]